### PR TITLE
Refs #30254 -- Added tests for Model.__hash__() inheritance.

### PR DESCRIPTION
**Pull request by carltongibson**
See https://code.djangoproject.com/ticket/30333 for more context. 

These tests pass but the first failed on 2.1. The change was a bug fix around #30254, rather than a regression. They serve as demonstrating the correct behaviour.

### CODE REVIEW

File: tests/basic/tests.py [hunk 1]
```diff
@@ -357,6 +357,24 @@ def test_hash(self):
             # hash)
             hash(Article())
 
+    def test_missing_hash_not_inherited(self):
+        class NoHash(models.Model):
+            def __eq__(self, other):
+                return super.__eq__(other)
+
+        with self.assertRaisesMessage(TypeError, "unhashable type: 'NoHash'"):
+            hash(NoHash(id=1))
```
Comment (felixxm) [inline tests/basic/tests.py]:
```suggestion
            hash(NoHash())
```

File: tests/basic/tests.py [hunk 2]
```diff
@@ -357,6 +357,24 @@ def test_hash(self):
...
+
+    def test_specified_parent_hash_inherited(self):
+        class ParentHash(models.Model):
+            def __eq__(self, other):
+                return super.__eq__(other)
+
+            __hash__ = models.Model.__hash__
+
+        hash(ParentHash(id=1))
```
Comment (felixxm) [inline tests/basic/tests.py]:
```suggestion
        self.assertEqual(hash(ParentHash(id=1)), 1)
```

File: tests/basic/tests.py [hunk 3]
```diff
@@ -357,6 +357,24 @@ def test_hash(self):
...
+        self.assertIs(ParentHash.__hash__, models.Model.__hash__)
```
Comment (felixxm) [inline tests/basic/tests.py]: I'm not sure if this assertion is necessary :thinking:.

File: tests/basic/tests.py (hunk 1 above)
Comment (carltongibson) [reply tests/basic/tests.py]: Models without ids can’t be hashed, which does raise a different error, but with the id is the minimum valid instance that can be hashed. 

As such, I put it in. 

Happy to go your way if you still prefer?

File: tests/basic/tests.py (hunk 2 above)
Comment (carltongibson) [reply tests/basic/tests.py]: What I really wanted was assert didn’t raise...

File: tests/basic/tests.py (hunk 3 above)
Comment (carltongibson) [reply tests/basic/tests.py]: Yeah. This whole case isn’t really necessary but it shows the correct usage and the correct result. 

Half way through __new__ (the old way) we get a new class which has the parent’s __eq__, but which later has its own. 

So... I put it to assert that we get the reference we’re actually expecting. It looks obvious, but it’s been wrong for years. 

(The behaviour was changed in Python 2.6, I learnt...)

File: tests/basic/tests.py (hunk 1 above)
Comment (felixxm) [reply tests/basic/tests.py]: Thanks for explanation, I just want to minimize test, so your version with `id` looks good :+1:

File: tests/basic/tests.py (hunk 3 above)
Comment (carltongibson) [reply tests/basic/tests.py]: This was the thread. It was quite interesting https://bugs.python.org/issue2235

**Comment by carltongibson** [comment]
Thanks for the comments. I’ve left replies. Happy to go whichever way you think. Let me know. I’ll finish over it the next few days.

**Comment by carltongibson** [comment]
@felixxm Updated as per your review. I'll just wait for CI and then pull it in.

File: tests/basic/tests.py [hunk 4]
```diff
@@ -357,6 +357,23 @@ def test_hash(self):
             # hash)
             hash(Article())
 
+    def test_missing_hash_not_inherited(self):
+        class NoHash(models.Model):
+            def __eq__(self, other):
+                return super.__eq__(other)
```
Comment (etanttila) [inline tests/basic/tests.py]: Maybe this should be just `pass` as `super.__eq__` shouldn't work (`super().__eq__` would). This is of course minor, as the line is never executed.

**Pull request merged**